3.workload.md

```diff
@@ -0,0 +1,378 @@
+| `spec`| [`Spec`](#spec) | Y | | A container for the workload definition. |
+
+### Metadata
+
+The metadata describes this workload definition. The `metadata` section is defined in [the meta section of part 2](2.overview_and_terminology.md#Metadata)
+
+### Spec
+
+The specification section contains a reference to the workload schema
+
+| Attribute | Type | Required | Default Value | Description |
+|-----------|------|----------|---------------|-------------|
+| `DefinitionRef` | [`DefinitionRef`](#DefinitionRef) | Y | | Workload schema reference. |
+
+
+#### DefinitionRef
+
+DefinitionRef defines where to find the workload schema
+
+| Attribute | Type | Required | Default Value | Description |
+|-----------|------|----------|---------------|-------------|
+| `name` | `string` | Y | | The workload schema's name. Must be formatted in a way that can uniquely identify a specific version of the definition  |
+
+The workload definition itself __MUST__ contain information that can be used to locate it directly. We will use that
+ information in the [Component](4.component.md) to refer to the workload schema instead of using the name of the workload definition. 
+
+We recommend that the `name` here follows the format described in [the Group, Version, Kind section of part 2](2.overview_and_terminology.md).
+
+We also recommend that the `name` of the workload (in the metadata section) to include the Group and Kind of the schema.
+
+## Workload Kinds
+
+There are three kinds of workloads:
+
+__Core workloads__ are the kind of workloads that every OAM runtime MUST implement according to spec definition
+
+__Standard workloads__ are the kind of workloads that a OAM runtime can choose to implement. However, the
+ implementation MUST strictly adhere to the schematics defined in the spec 
+
+__Extended workloads__ OAM platforms are free to define extra workload kinds and  
+
+### Core Workload Kind
+
+A core workload MUST be in the `core.oam.dev` group.
+
+A core workload MUST all be supported according to the definitions in this section by any runtime implementation of
+ this specification.
+
+Currently, all core workloads are container-based, and assume that an implementation is capable of executing an OCI or Docker image as a container, and are capable of working with OCI registries.
+We may introduce non container-based workloads, such as VM-based, in the future. We can also promote standard workloads to core if there is a consensus in the community.
+
+### Standard Workload Kind
+
+A standard workload MUST be in the `standard.oam.dev` group.
+
+A standard workload MUST be implemented according to the definitions in this section if any runtime chooses to implement it.
+ 
+We do not have any standard workload defined in the current spec. We may introduce new workloads or promote extended workloads to standard if there is a consensus in the community.
+ 
+### Extended Workload Kind
+An extended workload MUST be in the `extended.oam.dev` group.
```

Comment:
Part of our goal with this spec update was to allow arbitrary Kubernetes CRs to act as extended workloads. This isn't possible if all extended workloads MUST be in the `extended.oam.dev` API group. I don't think we should have an opinion about the group name of extended workloads.

Reply:
>  This isn't possible if all extended workloads MUST be in the `extended.oam.dev` API group.

Since the CRDs' group name does not have to be the same as its parent workloadDefinition's group name, I wonder why it is not possible to refer to any CRDs?

Reply:
Perhaps I'm misunderstanding something here. This reads to me like you're saying any extended workload, i.e. any kind blessed as a workload via a `WorkloadDefinition` must be in the `extended.oam.dev` group. So for example I could not use a Crossplane `RedisCluster` as a workload, because it's in `cache.crossplane.io`, not `extended.oam.dev`:

```yaml
apiVersion: core.oam.dev/v1alpha2
kind: Component
metadata:
  name: example-cache
spec:
  workload:
    # Are you saying this MUST be extended.oam.dev?
    apiVersion: cache.crossplane.io/v1alpha1
    kind: RedisCluster
    spec:
      engineVersion: "3.2"
```

Reply:
@negz  I think Ryan means the WorkloadDefinition under `extended.oam.dev` group instead of `core.oam.dev`? It should be mentioned in xxxDefinition section.

Reply:
Yeah, the description here has some mislead.  It is workloadDefinition MUST be in the `extended.oam.dev` group not workload itself.

Reply:
Does this mean we'll have multiple `kind: WorkloadDefinition`, each in a different API group (`extended.oam.dev`, `core.oam.dev`, etc?). I imagine core workloads will still require a `WorkloadDefinition`.

Reply:
Sorry @negz ,  I'm wrong.  I think we didn't need `extended.oam.dev` group at all.  I think the group `core.oam.dev` means they are absolutely defined in OAM Spec and every OAM runtime MUST follow it.  The group `standard.oam.dev` means they are defined in OAM Spec and can be optionally implemented by runtime.

Reply:
This also contradicts the recommendation above, which says that we recommend the workload definition name match the schema name. For extended workload types, that will be impossible if someone is adding existing workloads, like the RedisCluster in @negz 's example. 

I don't think we need the `extended.oam.dev` group at all here.

Reply:
> ```yaml
>   workload:
>     # Are you saying this MUST be extended.oam.dev?
>     apiVersion: cache.crossplane.io/v1alpha1
>     kind: RedisCluster
>     spec:
>       engineVersion: "3.2"
> ```
No, this is the apiVersion of the CRD. The extended workloadDefinition itself needs to be in the "extended.oam.dev". The CRD it refers to can be anything.

Reply:
> workload:
>     # Are you saying this MUST be extended.oam.dev?
>     apiVersion: cache.crossplane.io/v1alpha1
>     kind: RedisCluster
>     spec:
>       engineVersion: "3.2"
> 
> No, this is the apiVersion of the CRD. The extended workloadDefinition itself needs to be in the "extended.oam.dev". The CRD it refers to can be anything.

meaning
```
apiVersion: **extended**.oam.dev/v1alpha1
kind: WorkloadDefinition
metadata:
  name: rediscluster.cache.crossplane.io
spec:
  definitionRef:
    name: rediscluster.cache.crossplane.io
```
?

Reply:
> apiVersion: **extended**.oam.dev/v1alpha1
> kind: WorkloadDefinition
> metadata:
>   name: rediscluster.cache.crossplane.io
> spec:
>   definitionRef:
>     name: rediscluster.cache.crossplane.io

yeah,  is this ok?

Reply:
> allow arbitrary Kubernetes CRs to act as extended workloads.

I totally agree on this. OAM is supposed to manage, not replace CRDs. The "extended" needs some context that needs some explanation:

When we started the project, our vision is to standardize application management. There are levels of standardization in our roadmap: _core_ for minimum, _standard_ for cloud ubiquitous, and _extended_ for things that are popular but not ubiquitous. We host and govern those features/CRDs/Operators in OAM to build a standardized application delivery experience.

This is not a technical issue, but more a community building effort.

Reply:
Allow me summarize a bit, so Ryan proposed is:

For core workloads, it looks like:
```yaml
apiVersion: core.oam.dev/v1alpha1
kind: WorkloadDefinition
metadata:
  name: x
spec:
  definitionRef:
    name: x.core.oam.dev
```
For standard workload, it will be:

```yaml
apiVersion: standard.oam.dev/v1alpha1
kind: WorkloadDefinition
metadata:
  name: y.database.alibaba.com
spec:
  definitionRef:
    name: y.database.alibaba.com
```

For extended workload, it will be:

```yaml
apiVersion: extended.oam.dev/v1alpha1
kind: WorkloadDefinition
metadata:
  name: z.cache.crossplane.io
spec:
  definitionRef:
    name: z.cache.crossplane.io
```
Am I right?

Reply:
Here is what I think makes sense.  WorkloadDefinition always belongs to core.oam.dev as it's OAM's core concept. The apiground in the schema for core workload is core.oam.dev.  The apiground in the schema for standard workload is standard.oam.dev. There is no restriction on the api-group for extended.

For core workloads, it looks like:
```yaml
apiVersion: core.oam.dev/v1alpha2
kind: WorkloadDefinition
metadata:
  name: x.core.oam.dev
spec:
  definitionRef:
    name: x.core.oam.dev
```
For standard workload, it will be:
```yaml
apiVersion: core.oam.dev/v1alpha2
kind: WorkloadDefinition
metadata:
  name: y.standard.oam.dev
spec:
  definitionRef:
    name: y.standard.oam.dev
```
For extended workload, it will be:
```yaml
apiVersion: core.oam.dev/v1alpha2
kind: WorkloadDefinition
metadata:
  name: z.cache.crossplane.io
spec:
  definitionRef:
    name: z.cache.crossplane.io
```

Reply:
I think I had a stale view of this thread when I wrote https://github.com/oam-dev/spec/pull/316#discussion_r384223831. It sounds like I agree with @ryanzhang-oss's comment at https://github.com/oam-dev/spec/pull/316#discussion_r384219536.